tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -0,0 +1,336 @@
+
+from tensorflow.python.ops import gen_nn_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import variable_scope
+from tensorflow.python.ops import variables
+from tensorflow.python.training import optimizer
+from tensorflow.python.training import session_run_hook
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import data_flow_ops
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import constant_op
+
+
+class ElasticAverageCustomGetter(object):
+  """Custom_getter class is used to do:
+  1. Change trainable variables to local collection and place them at worker
+    device
+  2. Generate global variables(global center variables)
+  3. Generate local variables(local center variables) which record the global
+    variables and place them at worker device
+  4. Remain global step variable as global variable
+    Notice that the class should be used with tf.replica_device_setter,
+    so that the global center variables and global step variable can be placed
+    at ps device. Besides, use 'tf.get_variable' instead of 'tf.Variable' to
+    use this custom getter.
+
+  For example,
+  ea_custom_getter = ElasticAverageCustomGetter(worker_device)
+  with tf.device(
+    tf.train.replica_device_setter(
+      worker_device=worker_device,
+      ps_device="/job:ps/cpu:0",
+      cluster=cluster)),
+    tf.variable_scope('',custom_getter=ea_custom_getter):
+    hid_w = tf.get_variable(
+      initializer=tf.truncated_normal(
+          [IMAGE_PIXELS * IMAGE_PIXELS, FLAGS.hidden_units],
+          stddev=1.0 / IMAGE_PIXELS),
+      name="hid_w")
+    hid_b = tf.get_variable(initializer=tf.zeros([FLAGS.hidden_units]),
+                            name="hid_b")
+  """
+
+  def __init__(self, worker_device):
+    """Create a new `ElasticAverageCustomGetter`.
+
+    Args:
+      worker_device: String.  Name of the `worker` job.
+    """
+    self._worker_device = worker_device
+
+  def __call__(self, getter, name, *args, **kwargs):
+    if name.find('global_step') != -1:
+      return getter(name, *args, **kwargs)
+
+    kwargs['collections'] = [ops.GraphKeys.LOCAL_VARIABLES]
+    with ops.device(self._worker_device):
+      local_var = getter(name, *args, **kwargs)
+
+    if kwargs['trainable']:
+      global_center_variable = variables.Variable(
+        name='%s/%s' %
+             ('global_center_variable',
+              name),
+        initial_value=local_var.initialized_value(),
+        trainable=False,
+        collections=[
+          ops.GraphKeys.GLOBAL_VARIABLES,
+          'global_center_variable'])
+
+      with ops.device(self._worker_device):
+        local_center_variable = variables.Variable(
+          name='%s/%s' % ('local_center_variable', name),
+          initial_value=local_var.initialized_value(),
+          trainable=False,
+          collections=[ops.GraphKeys.LOCAL_VARIABLES,
+                       'local_center_variable'])
```

Comment:
Same thing for this collection name.

I wonder though if you're not better off using explicit data structures to communicate information from the getter to the optimizer. Maybe if you pass the getter as a constructor argument to the optimizer you can do that?

Reply:
I do not think passing the getter as a constructor to the optimizer will work. The local_var should be used to in forward computation and backward computation. But the global variables will be used in forward pass if the getter not used before building the model(getter returns the local_var).

Reply:
What I meant is that now you're using hidden global state in the optimizer to look up what is the global variable for a given local variable. Why not store a map in the custom_getter and make this map directly available to the optimizer by passing the custom_getter to the optimizer?

This way you do not need any new variable collections at all.